lib/ts/recipe/oauth2/types.ts

```diff
@@ -0,0 +1,179 @@
+/* Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ * This software is licensed under the Apache License, Version 2.0 (the
+ * "License") as published by the Apache Software Foundation.
+ *
+ * You may not use this file except in compliance with the License. You may
+ * obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { BaseRequest, BaseResponse } from "../../framework";
+import OverrideableBuilder from "supertokens-js-override";
+import { GeneralErrorResponse, JSONObject } from "../../types";
+
+export type TypeInput = {
+    override?: {
+        functions?: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis?: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeNormalisedInput = {
+    override: {
+        functions: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type APIOptions = {
+    recipeImplementation: RecipeInterface;
+    config: TypeNormalisedInput;
+    recipeId: string;
+    isInServerlessEnv: boolean;
+    req: BaseRequest;
+    res: BaseResponse;
+};
+
+export type RecipeInterface = {
+    tokensPOST(input: {
+        payload?: any;
+        validitySeconds?: number;
+        useStaticSigningKey?: boolean;
+        userContext: any;
+    }): Promise<
+        | {
+              token_type: "Bearer";
+              access_token: string;
+              refresh_token?: string;
+              id_token?: string;
+              expires_in: number;
+          }
+        | GeneralErrorResponse
+    >;
+    // The OpenId spec states that we have to support both POST and GET
+    authorizeGET(input: {
+        clientId: string;
+        responseType: string;
+        redirectUri?: string;
+        scopes: string[];
+        state?: string;
+
+        // PKCE
+        codeChallenge?: string;
+        codeChallengeMethod?: string;
+
+        // OpenIdConnect
+        prompt?: string;
+        nonce?: string;
+
+        // Not sure if we want to support these
+        acr_values?: string;
+        maxAge?: number;
+
+        options: APIOptions;
```

Comment:
missing param for `mode` (online or offline)

Reply:
Not missing from here. Offline access is requested through the `offline_access` according to the OId specs, and I didn't see a way to specifically request it in base OAuth2. The basic idea was that we'd choose if we should issue a refresh token in the BE SDK (in some overrideable function, i.e.: `shouldIssueRefreshToken`).
We could remove the related param from CDI and rely on scopes in the core as well, but this seemed like a more flexible solution.

Reply:
Need to discuss this. Im not sure i understand